dali/python/nvidia/dali/worker.py

```diff
@@ -0,0 +1,159 @@
+# Copyright (c) 2020, NVIDIA CORPORATION. All rights reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import threading
+from multiprocessing import reduction
+from nvidia.dali.shared_batch import SharedMemChunk, SharedBatchSerialized, SharedBatchWriter
+
+
+class SharedBatchesDispatcher:
+    """SharedBatchesDispatcher serializes batches, puts them into provided
+shared memory chunks and notifies parent process of batch ready to be read from shared memory.
+It keeps track of what shared memory chunks have been already sent and if needed, sends
+file descriptors of memory chunks that parent process hasn't seen yet.
+
+Parameters
+----------
+`worker_id` : int
+    Id of the worker passed by the parent process. Added to messages sent over the ``res_pipe``
+    to simplify bookeeping in parent process.
+`sock` : socket
+    Python wrapper around Unix socket, capable of sending file descriptors between processes.
+`res_pipe`: pipe
+    Pipe used to send parent process a notification (along with essential meta data info) about
+    ready batch in a given shared memory chunk.
+"""
+    def __init__(self, worker_id, sock, res_pipe):
+        self.worker_id = worker_id
+        self.fd_sent = set()
+        self.sock = sock
+        self.res_pipe = res_pipe
+
+    def send(self, mem_chunk, batch_data, header):
+        if mem_chunk is None:
+            self.res_pipe.send((self.worker_id, header, batch_data))
```

Comment:
So we can pass the batch by value or using mem_chunk?

Reply:
Using mem_chunk. I'll leave there a comment - ``mem_chunk is None`` is a special case when StopIteration was raised (and I am changing it to handle any exception), in that case batch is not written to shared memory and worker notifies simply the main process about the exception through the pipe.